E2E: add new spec to test the WordPress.com Free plan.

worldomonation opened the pull request:
#### Changes proposed in this Pull Request

This PR adds new methods and spec to test the WordPress.com Free plan.

Key changes:
1. addition of new WP REST API methods.
2. addition of new spec file to teset the WordPress.com Free plan.

Details:
1. new method are added to interact with the WP REST API with the ultimate goal to support future specs. New methods added include interaction with the following endpoints:
  - /me
  - /me/account/close

2. this new spec corresponds to the `Plans: Sign up for Free plan` entry in the [project thread](pciE2j-Uk-p2), with a minor adjustment to use the UI for the signup process.

The rough outline of the spec is as follows:
- create a new account
- select the free plan
- navigate directly to the dashboard, ignoring any onboarding flows
- verify the plans grid under /plans
- delete the account

#### Testing instructions

Trigger the `Pre-Release` tests manually and verify that this spec does not fail.

Related to #

worldomonation committed:
- re-implement API interaction wtih node-fetch instead of got.
- implement RestAPIClient class with initial methods to interact with the API.

Revise the RestAPIClient class based on feedback.
- use smaller methods to build up the request params.
- remove use of GenericDict.
- add JSDOc where missing.

worldomonation committed:
Add new spec Plans: Signup free
- create new spec to sign up as new user and opt for the free plan.
- implement API method interactions to delete the user account.

worldomonation committed:
Move bearer token method out of TestAccounts.

Change the `getBearerToken` method to async and make it private.

worldomonation committed:
Rebase and revise the changes:
- fill out missing JSDoc
- rename the test suite

worldomonation commented on packages/calypso-e2e/src/lib/pages/user-signup-page.ts:
```diff
@@ -63,10 +69,18 @@ export class UserSignupPage {
+			throw new Error( 'Failed to create new user at signup.' );
+		}
+
+		const responseBody: NewUserResponse = JSON.parse( ( await response.body() ).toString() );
+		return { ID: responseBody.body.user_id, bearer_token: responseBody.body.bearer_token };
```
This is similar to what we do in `EditorPage.publish`. By intercepting and decoding the response when we create a new user (via signup), the data can then be used to cross-reference that we are operating on the expected user when interacting with the API.

worldomonation commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -16,21 +16,43 @@ interface RequestParams {
+interface AllSitesResponse {
 	sites: Site[];
 }
+interface CalypsoPreferencesResponse {
+	calypso_preferences: {
+		recentSites: number[];
+	};
+}
+interface MyAccountInformationResponse {
+	ID: number;
+	username: string;
+	email: string;
+}
+export interface NewUserResponse {
+	code: number;
+	body: {
+		success: boolean;
+		user_id: number;
+		username: string;
+		bearer_token: string;
+	};
+}
```
At some point I may have to move these into its own Types file.

dpasque replied:
Totally fair! I think they're fine here for now FWIW 😄

worldomonation commented on test/e2e/specs/plans/plans__signup-free.ts:
```diff
@@ -0,0 +1,119 @@
+declare const browser: Browser;
+
+describe(
+	DataHelper.createSuiteTitle( 'Plans: Create a WordPress.com Free site as a new user' ),
```
This suite is very similar to the older one - `signup__free.ts`. The difference between the two being the older suite meanders through the onboarding flow some more. 

Also note that the older suite is in the Quarantined build and while it's being run, it isn't gatekeeping the release.

I can think of several options:

1. remove the older suite in favor of the new one.
2. keep both suites and differentiate them some more.
3. truncate this spec to be more minimal.

For option 3, truncation means using an existing user instead and not performing much (if any) verification.

What do you think?

dpasque replied:
>Also note that the older suite is in the Quarantined build and while it's being run, it isn't gatekeeping the release.

This appears to [not be true actually](https://github.com/Automattic/wp-calypso/blob/trunk/test/e2e/specs/onboarding/signup__free.ts#L2)! And I think the existing test is useful because it is testing the First Ten Minutes/Hero Flow.  We definitely should continue to have coverage for that flow, especially as we roll out verticals, etc.

I think for now, we just leave the older suite as it is. We should come back around and probably redesign our E2E tests for the First Ten Minutes flow. This test can remain focused on account creation and the details of the free plan, where other tests are more focused on that onboarding flow and its consequences.

worldomonation replied:
OK, so it sounds like you're in favor of keeping both specs for now. I think that's acceptable - I will make sure to keep an eye on the test account signup numbers to make sure we stay below some arbitrary number.

dpasque commented on test/e2e/specs/plans/plans__signup-free.ts:
```diff
@@ -0,0 +1,119 @@
+			const response = await restAPIClient.closeAccount( userID, username, email );
+
+			if ( response.success !== true ) {
+				console.log( `Failed to delete user ID ${ userID }` );
```
Do we want to make this stand out as a warning more obviously? Maybe use `console.warn`, or make the text be a bit louder? Just so we notice it and can follow-up on it?

worldomonation replied:
I think that would make sense. We don't want to fail the whole build if this process fails, but you're right that it should stand out some more.

dpasque commented on packages/calypso-e2e/src/rest-api-client.ts:
```diff
@@ -147,11 +191,78 @@ export class RestAPIClient {
+	 * @returns {Promise<boolean>} True if account closure was successful. False otherwise.
+	 */
+	async closeAccount(
+		userID: number,
+		username: string,
+		email: string
```
I love the validation we do with this! We are asking consumers to essentially double-check the details of the account they expect to be deleted.

Small, optional, stylistic comment: can we convert all these parameters to properties on a single object parameter, maybe called `expectedAccountDetails`? Right now, looking at just the signature, it's confusing why we need this information when calling (because the bearer token and the class constructor should contain all the account-specific information.) Having it all in one descriptively named object might make it a bit more clear to consumers that they're providing "expected" account information for a double-check.

worldomonation replied:
> can we convert all these parameters to properties on a single object parameter,

I agree. Will make the change in the next commit.

dpasque commented on packages/calypso-e2e/src/lib/components/sidebar-component.ts:
```diff
@@ -12,6 +12,7 @@ const selectors = {
 
 	// Buttons and links within Sidebar
 	linkWithText: ( text: string ) => `a:has-text("${ text }")`,
+	planName: '.sidebar__inline-text',
```
This works for now, but I'm a little hesitant about relying on only a stylistic class name. Could we perhaps position this within the "Upgrades" row?

Like `.sidebar__menu:has-text("Upgrades") .sidebar__inline-text`

worldomonation replied:
That's fair. I'll use the same selector I do in other part of `SidebarComponent` to narrow it down a bit more.

dpasque reviewed (approved):
Great work on this one! That Rest API class is looking slick. 😎 

Left a few optional suggestions for your consideration. Otherwise, 🚀

worldomonation committed:
Update based on comments left on PR:
- add some more comments around the closeAccount flow.
- turn the parameters of `closeAccount` into an object.
- use console.warn if account deletion fails.
- narrow selector for plan name on the sidebar.

worldomonation committed:
Fix test lint

worldomonation committed:
Revert changes made in #63099 that changed the plan names from "Free" to "start
with Free".
Change the methods in RestAPIClient.closeAccount to send `console.warn` instead of
throwing.

worldomonation committed:
Fix test eslint
